common/protos/proto/metadata.proto

```diff
@@ -0,0 +1,248 @@
+message TableIdent {
+  uint64 ver = 100;
+
+  uint64 table_id = 1;
+
+  // The seq of TableMeta
+  uint64 seq = 2;
+}
+
+// TableMeta is a container of all non-identity information.
+message TableMeta {
+  uint64 ver = 100;
+
+  // Table schema, i.e., columns definition.
+  DataSchema schema = 1;
+
+  // Table engine.
+  string engine = 6;
+
+  // Database engine options, like github engine <token>.
+  map<string, string> engine_options = 7;
+
+  // Table options.
+  map<string, string> options = 5;
+
+  // Keys to sort rows in table.
+  optional bytes order_keys = 9;
+
+  // The time table created.
+  string created_on = 20;
+}
+
+// The schema of a table, such as column data types and other meta info.
+message DataSchema {
+  uint64 ver = 100;
+
+  // Fields in the table
+  repeated DataField fields = 1;
+
+  // Other optional table info
+  map<string, string> metadata = 2;
+}
+
+// One field, AKA column
+message DataField {
+  uint64 ver = 100;
+
+  // The name of this column.
+  string name = 1;
+
+  // A SQL style expression to define the default value for this field.
+  optional bytes default_expr = 2;
+
+  // Column data type
+  DataType data_type = 3;
+}
+
+// An enumeration of all supported data types.
+message DataType {
+  uint64 ver = 100;
+
+  oneof dt {
+    NullableType nullable_type  = 1;
+    Empty        bool_type      = 2;
+    Empty        int8_type      = 3;
+    Empty        int16_type     = 4;
+    Empty        int32_type     = 5;
+    Empty        int64_type     = 6;
+    Empty        uint8_type     = 7;
+    Empty        uint16_type    = 8;
+    Empty        uint32_type    = 9;
+    Empty        uint64_type    = 10;
+    Empty        float32_type   = 11;
+    Empty        float64_type   = 12;
+    Empty        date_type      = 13;
+    Timestamp    timestamp_type = 14;
+    Empty        string_type    = 15;
+    Struct       struct_type    = 16;
+    Array        array_type     = 17;
+    Variant      variant_type   = 18;
+  }
+}
+
+// A wrapper data type of another type.
+// Such a column allows to contain `null` elements
+message NullableType {
+  uint64 ver = 100;
+
+  // The type for the non-null element.
+  DataType inner = 1;
+}
+
+// Timestamp data type with customizable precision and timezone: `tz`.
+message Timestamp {
+  uint64 ver = 100;
+
+  /// The time resolution is determined by the precision parameter, range from 0 to 9
+  /// Typically are used - 0 (seconds) 3 (milliseconds), 6 (microseconds), 9 (nanoseconds).
+  uint64 precision = 1;
+
+  /// tz indicates the timezone, if it's None, it's UTC.
+  optional string tz = 2;
+}
+
+// Struct is similar to a `map` with fixed keys.
+message Struct {
+  uint64 ver = 100;
+
+  // Struct field names.
+  repeated string names = 1;
+
+  // The corresponding field types.
+  repeated DataType types = 2;
+}
+
+// Array contains multiple elements of the same type.
+message Array {
+  uint64 ver = 100;
+
+  // The type of the elements
+  DataType inner = 1;
```

Comment:
`inner` element in `Array` seems to be repeated?

Reply:
If you mean recursively nested, yes.

If you mean whether there will multiple element types in an array. IMHO it is not possible

Reply:
emm, i guess `inner` is container of array elements, but actually it is array type.